contracts/cw20-merkle-airdrop/src/contract.rs

```diff
@@ -0,0 +1,486 @@
+
+    Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::UpdateConfig { owner } => execute_update_config(deps, env, info, owner),
+        ExecuteMsg::RegisterMerkleRoot { merkle_root } => {
+            execute_register_merkle_root(deps, env, info, merkle_root)
+        }
+        ExecuteMsg::Claim {
+            stage,
+            amount,
+            proof,
+        } => execute_claim(deps, env, info, stage, amount, proof),
+    }
+}
+
+pub fn execute_update_config(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    owner: Option<String>,
+) -> Result<Response, ContractError> {
+    // authorize owner
+    let cfg = CONFIG.load(deps.storage)?;
+    if info.sender != cfg.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // validate owner change
+    let new_owner = owner.ok_or(ContractError::InvalidInput {})?;
+    let new_owner = deps.api.addr_validate(&new_owner)?;
+
+    CONFIG.update(deps.storage, |mut exists| -> StdResult<_> {
+        exists.owner = new_owner;
+        Ok(exists)
+    })?;
+
+    Ok(Response {
+        messages: vec![],
+        attributes: vec![attr("action", "update_config")],
+        data: None,
+        events: vec![],
+    })
+}
+
+pub fn execute_register_merkle_root(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    merkle_root: String,
+) -> Result<Response, ContractError> {
+    // authorize owner
+    let cfg = CONFIG.load(deps.storage)?;
+    if info.sender != cfg.owner {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // check merkle root length
+    let mut root_buf: [u8; 32] = [0; 32];
+    hex::decode_to_slice(merkle_root.to_string(), &mut root_buf)?;
+
+    let stage = STAGE.update(deps.storage, |stage| Ok(stage + 1))?;
+
+    MERKLE_ROOT.save(deps.storage, U8Key::from(stage), &merkle_root)?;
+    STAGE.save(deps.storage, &stage)?;
+
+    Ok(Response {
+        messages: vec![],
+        attributes: vec![
+            attr("action", "register_merkle_root"),
+            attr("stage", stage),
+            attr("merkle_root", merkle_root),
+        ],
+        data: None,
+        events: vec![],
+    })
+}
+
+pub fn execute_claim(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    stage: u8,
+    amount: Uint128,
+    proof: Vec<String>,
+) -> Result<Response, ContractError> {
+    // verify not claimed
+    let claimed = CLAIM.may_load(deps.storage, (&info.sender, U8Key::from(stage)))?;
+    if claimed.is_some() {
+        return Err(ContractError::Claimed {});
+    }
+
+    let config = CONFIG.load(deps.storage)?;
+    let merkle_root = MERKLE_ROOT.load(deps.storage, stage.into())?;
+
+    let user_input = format!("{}{}", info.sender, amount);
+    let hash = sha3::Keccak256::digest(user_input.as_bytes())
```

Comment:
These lines (142/143-157) should be pulled out into a separate function to heavily unit test it and compare with TS test vectors. They are simple function on basic data structures and easier to heavily test than the execute methods